### ISSUE TITLE
Consistency between source code and published papers

### ISSUE THREAD

**Issue by DongxingHE667** [NaChuaN666/UP-CrackNet#8]
The source code performs additional per-image min–max renormalization on the input image during the testing phase and changes the numerical domain, resulting in the residual calculation no longer strictly corresponding to the paper formula (1).The original pixel spatial difference is inconsistent with the test process in the paper.Why?I hope you can take time out of your busy schedule to reply. Thank you very much.

**Comment by NaChuaN666** [comment]
Thanks for pointing this out. I'll check it this week. Appreciate your careful review.

**Comment by NaChuaN666** [comment]
Hi Dongxing, Thank you for your careful review. The issue you pointed out is indeed an error in our code. During the testing phase, we performed an extra min-max renormalization on the input and generated images, which caused the error map calculation to deviate from the formula in the paper. 
You can update the code in the utilis.py as:

def save_error_maps_gray_corrected(input_name, label, input_ori, target, gen_image,
                                  epoch, save=False, save_dir='output_results/'):

    name = input_name.split(".")[0]
   
    input_vis = (input_ori[0] + 1) / 2
    gen_vis = (gen_image[0] + 1) / 2
    
    input_np = input_vis.numpy().transpose(1, 2, 0)
    gen_np = gen_vis.numpy().transpose(1, 2, 0)
    
    input_gray = cv.cvtColor(input_np, cv.COLOR_RGB2GRAY)
    gen_gray = cv.cvtColor(gen_np, cv.COLOR_RGB2GRAY)
    
    error_map = np.abs(input_gray - gen_gray)
    error_map_uint8 = (error_map * 255).astype(np.uint8)
    if save:
        if not os.path.exists(save_dir):
            os.mkdir(save_dir)
        cv.imwrite(save_dir + '{:s}.png'.format(name), error_map_uint8)

In addition, we re-ran experiments using the uploaded checkpoint. With the bilateral filter parameters set to cv.bilateralFilter(img, 38, 275, 38), we achieved an IoU of 43.566%, which is slightly lower than values in the published paper. Interestingly, this bug seems to have had a beneficial effect for road crack detection, possibly because the extra normalization inadvertently enhanced the contrast of low-contrast crack regions. This observation may inspire further investigation into preprocessing strategies for crack detection. Thank you again for your time and valuable contribution to improving our work.

**Comment by DongxingHE667** [comment]
How are the last three parameters in the function cv.bilateralFilter determined? Is it determined based on experience? Because I found that the version published in github is different from the one used this time. I hope the author can reply, thank you very much.

**Comment by NaChuaN666** [comment]
Thank you for your question. The bilateral filter parameters were determined empirically through experiments on the specific dataset. We conducted a series of tests with different parameter combinations and selected the ones that achieved the best detection performance. The parameters in the original GitHub version were based on our initial experiments, while the updated ones were obtained after re-tuning following the bug fix. This adjustment helped partially recover the performance. In practice, these parameters may need to be re-tuned for different datasets.

**Comment by DongxingHE667** [comment]
Thank you for your detailed reply.In addition, I re-ran train.py, and it exited abnormally at the 51st epoch. The reason was Loss.cu: Assertion input_val >=0 && <=1 + device-side assert triggered. During the training process, the value input to BCELoss by the discriminator was not in the range [0,1] (or NaN was generated), triggering CUDA's assertion check, causing the program to exit abnormally. May I ask if the author has ever encountered this kind of situation?Make sure the code you upload is the latest version?

**Comment by NaChuaN666** [comment]
For reference, here's my environment that trained successfully:
PyTorch version: 1.11.0+cu113
CUDA version: 11.3
GPU: NVIDIA GeForce RTX 4090 D
This setup ran for 200 epochs without any issues. The version differences might explain the assert behavior you're seeing.

**Comment by DongxingHE667** [comment]
I checked and the versions of pytorch and cuda are consistent. Since the video memory of my computer is relatively small, I made the following adjustments: batchsize=4, num_workers=4, prefetch_factor=2. This shouldn't have any impact on training, right?

**Comment by NaChuaN666** [comment]
Your adjustments (batchsize=4, workers=4) are reasonable and should work fine. Since the error is about values outside [0,1], I suggest adding this check to catch potential data issues: 
if torch.isnan(input).any() or torch.isinf(input).any():
    print("NaN/Inf detected"); continue
This will help identify if the problem comes from data loading.

**Comment by DongxingHE667** [comment]
Thanks for your guidance. Now it can be trained normally. We made the following modifications based on what was published in github: epoch=200, num_workers=4, prefetch_factor=4. No adjustments were made to the rest. We were surprised to find that the IOU indicator dropped to 41.0041. Why is this?

**Comment by NaChuaN666** [comment]
1. Extending to 300 epochs may improve performance.
2. Bilateral filter parameters in testing: The IOU is also affected by the bilateral filter settings during testing. When using the corrected save_error_maps_gray_corrected function (without the bug), the original filter parameters (25, 450, 15) may no longer be optimal. We found that tuning them to (38, 275, 38) gave better results (43.56% IoU).
3. The bug's impact: Interestingly, when the original normalization bug was present, the model achieved 44.286% IoU with (25, 450, 15). This suggests the bug may act as an implicit contrast enhancement that benefited crack detection.

**Comment by DongxingHE667** [comment]
I would like to thank the author for taking the time to respond very patiently and comprehensively to my questions. Thank you again.
